src/components/modals/Receive/steps/04-step-receive-funds.js

```diff
@@ -4,55 +4,69 @@ import invariant from 'invariant'
 import React, { PureComponent } from 'react'
 
 import TrackPage from 'analytics/TrackPage'
+import getAddress from 'commands/getAddress'
+import { isSegwitAccount } from 'helpers/bip32'
 import Box from 'components/base/Box'
-import Label from 'components/base/Label'
 import CurrentAddressForAccount from 'components/CurrentAddressForAccount'
-import RequestAmount from 'components/RequestAmount'
+import { WrongDeviceForAccount } from 'components/EnsureDeviceApp'
 
-import type { StepProps } from '../index'
+import type { StepProps } from '..'
 
-type State = {
-  amount: number,
-}
+export default class StepReceiveFunds extends PureComponent<StepProps> {
+  componentDidMount() {
+    if (this.props.isAddressVerified === null) {
+      this.confirmAddress()
+    }
+  }
+
+  confirmAddress = async () => {
+    const { account, device, onChangeAddressVerified, transitionTo } = this.props
+    invariant(account, 'No account given')
+    invariant(device, 'No device given')
+    try {
+      const params = {
+        currencyId: account.currency.id,
+        devicePath: device.path,
+        path: account.freshAddressPath,
+        segwit: isSegwitAccount(account),
+        verify: true,
+      }
+      const { address } = await getAddress.send(params).toPromise()
 
-export default class StepReceiveFunds extends PureComponent<StepProps, State> {
-  state = {
-    amount: 0,
+      if (address !== account.freshAddress) {
+        throw new WrongDeviceForAccount(`WrongDeviceForAccount ${account.name}`, {
+          accountName: account.name,
+        })
+      }
+      onChangeAddressVerified(true)
+      transitionTo('receive')
+    } catch (err) {
+      onChangeAddressVerified(false, err)
+      this.props.transitionTo('confirm')
+    }
   }
 
-  handleChangeAmount = (amount: number) => this.setState({ amount })
   handleGoPrev = () => {
+    // FIXME this is not a good practice at all. it triggers tons of setState. these are even concurrent setState potentially in future React :o
```

Comment:
Yeah.. but didn't really find decent alternative for now :/ At first, parent passed its own `setState` method to `StepProps`, but it rapidly appears as weird design (even if it solves the pb). For now it works pretty well because used on small views.

Wdyt of `ReactDOM.unstable_batchedUpdates`? (relatable [facebook/react#10231](https://github.com/facebook/react/issues/10231#issuecomment-316644950))

Dyou think of better approach?

Reply:
I think we could just pass a prop that represent exactly what this action is about. exactly like on Redux, it would just "one action" and just "multiple reducers treat it". here it can be just one callback as prop, and just one setState that change what needs.

what fundamentally need to be reworked however, is to be able to "control" everything. for instance, the stepper does not allow to have its step controlled controlled. basically it shouldn't have a `transitionTo`, it should just be an internal detail that is called during componentDidUpdate and when stepId changes. 🤔

once this, we can control everything from the "top level" that handle the business logic (the modal body level) , top-down data flow basically and one state at same place.

Reply:
sounds good